### PR TITLE
Distributions serve RepositoryVersion

### PR DESCRIPTION
**Pull request by bmbouter**
This adds the field 'repository_version' to a Distribution which will
serve the content in that RepositoryVersion.

This also removes the 'publisher' field which was decided to no longer
be something we want at this time.

It adds some docs in the workflow section that describing 3
auto-distribution workflows.

Lastly, it adjusts auto-distribution so that the 'repository' field will
serve the latest RepositoryVersion automatically.

The 'repository', 'repository_version', and 'publication' fields cannot
be used together.

https://pulp.plan.io/issues/4745
closes #4745

### CODE REVIEW

**Comment by kersommoura** [comment]
@bmbouter, explained me that the auto-distribution has been updated. And new test cases will be filed to reflect those new changes. That is why some tests cases were removed.

**Comment by bmbouter** [comment]
I'm going to file several test issues today describing how we need to test this since this PR is removing tests. I'll link to them here on this issue as well later.

File: docs/workflows/exposing-content.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    updates to ``base_path`` are run serially by the tasking system.
+
+
+Auto-Distribution
+-----------------
+
+In this workflow you pair a :term:`Repository` and a :term:`Distribution` such that the Distribution
+will serve the latest RepositoryVersion associated with that Repository.
+
+First lets make a Repsotory named ``foo`` and save its URL as ``REPO_HREF``::
```
Comment (kersommoura) [inline docs/workflows/exposing-content.rst]:
```suggestion
First lets make a Repository named ``foo`` and save its URL as ``REPO_HREF``::
```

File: docs/workflows/exposing-content.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    documentation for details on if it uses a :term:`Publication` or not.
+
+
+Manual Distribution of a RepositoryVersion
+------------------------------------------
+
+In this workflow, you already have a :term:`RepositoryVersion` created. You then want to distribute
+its content at the base_path ``mypath`` using a :term:`Distribution`. In this case you manually
+associate the :term:`Distribution` with the :term:`RepositoryVersion` using the
+``repository_version`` option of the :term:`Distribtuion`.
```
Comment (kersommoura) [inline docs/workflows/exposing-content.rst]:
```suggestion
``repository_version`` option of the :term:`Distribution`.
```

File: docs/workflows/exposing-content.rst [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    documentation for details on if it uses a :term:`Publication` or not.
+
+
+Manual Distribution of a Publication
+------------------------------------
+
+In this workflow, you already have a :term:`Publication` created. You then want to distribute its
+content at the base_path ``mypath`` using a :term:`Distribution`. In this case you manually
+associate the :term:`Distribution` with the :term:`Publication` using the ``publication`` option of
+the :term:`Distribtuion`.
```
Comment (kersommoura) [inline docs/workflows/exposing-content.rst]:
```suggestion
the :term:`Distribution`.
```

File: docs/workflows/exposing-content.rst (hunk 1 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: will fix in next push, thank you!

File: docs/workflows/exposing-content.rst (hunk 2 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: fixed in next push, ty.

File: docs/workflows/exposing-content.rst (hunk 3 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: fixed in next push, ty

File: docs/glossary.rst [hunk 4]
```diff
@@ -14,12 +14,13 @@ Glossary
...
-        A `Django <https://docs.djangoproject.com>`_ app provided by :term:`pulpcore` that serves
-        :term:`content`.
+        An `aiohttp.server <https://aiohttp.readthedocs.io/en/stable/web.html>`_ app provided by
+        :term:`pulpcore` that serves :term:`content` through :term:`Distributions <distribution>`.
 
     :class:`~pulpcore.plugin.models.Distribution`
-        User facing settings that specify how and where associated
-        :term:`publications<publication>` are served.
+        User facing object that configures the :term:`content app` to serve either a
+        :term:`Repository Version <RepositoryVersion>`, a :term:`Repository`, or a
```
Comment (dkliban) [inline docs/glossary.rst]: Repository seems to be formatted differently than Repository Version and publication.

File: docs/workflows/exposing-content.rst [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+Overview
+--------
+
+Content, e.g. rpms or docker containers, loaded into Pulp is only served by Pulp if made available
+through a :term:`Distribution`. There are three workflows:
+
+* Auto-distribution of a RepositoryVersion
+* Manual distribution of a RepositoryVersion
+* Manual distribution of a Publication
+
```
Comment (asmacdo) [inline docs/workflows/exposing-content.rst]: It would be good if somewhere in here, it is explicit that plugins will typically use:

    (auto-distribution of repo version and/or manual distribution of repo version) OR (publications). 

This will add clarity that these are not so much "options for users" but "options for plugins". These docs should also instruct the user to see each plugin's docs for this, which should in turn refer to the workflows that work for that plugin.

Some of ^ might simplify out once we go to master/detail distributions though, so I'm fine with making that a separate issue, blocked by master/detail distributions.

File: docs/glossary.rst (hunk 4 above)
Comment (bmbouter) [reply docs/glossary.rst]: It's because I don't need to mutate `Repository` but I add a space to `RepositoryVersion` and an s to `Publication`.

File: docs/workflows/exposing-content.rst [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+``someexample`` or ``a/nested/example`` can be expected respectively::
+
+    /pulp/content/someexample/
+    /pulp/content/a/nested/example/
+
+
+.. note::
+
+    The ``base_path`` must have no overlapping components. So if a :term:`Distribution` with
+    ``base_path`` of ``a/path/foo`` existed, you could not make a second :term:`Distribution` with a
```
Comment (asmacdo) [inline docs/workflows/exposing-content.rst]: These examples suggest that the trailing and preceeding slashes are optional. Is that true?

File: docs/workflows/exposing-content.rst [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    `base_path` of ``a/path/bar`` or ``a/somethingelse`` because both contain subpaths of
+    ``a/path/foo``. Pulp will stop you from doing this which is why Distribution creates or
+    updates to ``base_path`` are run serially by the tasking system.
```
Comment (dkliban) [inline docs/workflows/exposing-content.rst]: this whole example is wrong. ``a/path/foo`` is not a subpath of ``a/path/bar`` or ``a/somethingelse``.

File: docs/workflows/exposing-content.rst [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    # Create a Publication
+    http POST :24817/pulp/api/v3/publications/file/file/ repository=$REPO_HREF publisher=$PUBLISHER_HREF
+    export PUBLICATION_HREF=$(http :24817/pulp/api/v3/publications/file/file/ | jq -r --arg PUBLISHER_HREF "$PUBLISHER_HREF" '.results[] | select(.publisher==$PUBLISHER_HREF) | ._href')
+
+Now with your :term:`Publication` saved as ``PUBLICATION_HREF`` you can have the
+:term:`Distribution` serve it at base_path ``bar``::
+
+    http POST :24817/pulp/api/v3/distributions/ name='baz' base_path='bar' publication=$PUBLICATION_HREF
+
+As soon as this is created, the :term:`Publication` will be immediately available at base_path
```
Comment (asmacdo) [inline docs/workflows/exposing-content.rst]: Optional: s/this is created/the generated task completes/

File: pulpcore/app/migrations/0001_initial.py [hunk 9]
```diff
@@ -329,9 +329,9 @@ class Migration(migrations.Migration):
                 ('base_path', models.CharField(max_length=255, unique=True)),
                 ('content_guard', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.ContentGuard')),
                 ('publication', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.Publication')),
-                ('publisher', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.Publisher')),
                 ('remote', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.Remote')),
                 ('repository', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.Repository')),
+                ('repository_version', models.ForeignKey(null=True, on_delete=django.db.models.deletion.SET_NULL, related_name='_distributions', to='core.RepositoryVersion')),
```
Comment (asmacdo) [inline pulpcore/app/migrations/0001_initial.py]: I would expect that if fields changed, the api.json should have also changed and should be committed here as well.

File: pulpcore/app/models/publication.py [hunk 10]
```diff
@@ -271,19 +243,15 @@ class Distribution(BaseDistribution):
     """
     A distribution defines how a publication is distributed by Pulp's webserver.
 
+    The `base_path` must have no overlapping components. So if a Distribution with `base_path` of
+    ``a/path/foo`` existed, you could not make a second Distribution with a `base_path` of
+    ``a/path/bar`` or ``a/somethingelse`` because both contain subpaths of ``a/path/foo``.
```
Comment (dkliban) [inline pulpcore/app/models/publication.py]: this is the same example as above and needs to be corrected also.

File: pulpcore/app/models/publication.py [hunk 11]
```diff
@@ -231,37 +204,36 @@ class ContentGuard(MasterModel):
 
 class BaseDistribution(Model):
     """
-    A distribution defines how a publication is distributed by Pulp's webserver.
+    A distribution defines how a publication is distributed by the Content App.
```
Comment (asmacdo) [inline pulpcore/app/models/publication.py]: Nice catch, thanks

File: pulpcore/app/serializers/publication.py [hunk 12]
```diff
@@ -104,6 +104,12 @@ class BaseDistributionSerializer(ModelSerializer):
     _href = IdentityField(
         view_name='distributions-detail'
     )
+    content_guard = DetailRelatedField(
```
Comment (asmacdo) [inline pulpcore/app/serializers/publication.py]: Is this change intended?

File: pulpcore/app/serializers/publication.py [hunk 13]
```diff
@@ -113,50 +119,60 @@ class BaseDistributionSerializer(ModelSerializer):
...
+    def validate(self, data):
+        super().validate(data)
+
+        mutex_keys = ['publication', 'repository', 'repository_version']
+        in_use_keys = []
+        for mkey in mutex_keys:
+            if mkey in data:
+                in_use_keys.append(mkey)
+
+        if len(in_use_keys) > 1:
```
Comment (asmacdo) [inline pulpcore/app/serializers/publication.py]: It is unclear in the comments and help_text that a Distribution is valid even if it has 0 of these keys.

**Comment by dkliban** [comment]
This change has no impact on pulp_maven. The pull-through cache functionality of pulp-content-app works correctly with this change.

File: docs/workflows/exposing-content.rst (hunk 5 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: I'll add a paragraph clarifying this some here and we can improve it more later too.

File: docs/workflows/exposing-content.rst (hunk 6 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: The trailing slashes are optional. For instance we currently use `dev` in many examples and it also has no trailing slash.

File: docs/workflows/exposing-content.rst (hunk 7 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: I will rewrite these now and include in the next push.

File: docs/workflows/exposing-content.rst (hunk 5 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: see if this next push is better.

File: docs/workflows/exposing-content.rst (hunk 8 above)
Comment (bmbouter) [reply docs/workflows/exposing-content.rst]: I opt for the existing language because I find it clearer for the reader and technically the object isn't created until the task runtime anyway. Ty for the suggestion anyway.

File: pulpcore/app/migrations/0001_initial.py (hunk 9 above)
Comment (bmbouter) [reply pulpcore/app/migrations/0001_initial.py]: oh shiz I gotta do this. Thank you.

Comment (bmbouter) [reply pulpcore/app/migrations/0001_initial.py]: will fix in next push.

File: pulpcore/app/models/publication.py (hunk 10 above)
Comment (bmbouter) [reply pulpcore/app/models/publication.py]: will fix in next push.

File: pulpcore/app/serializers/publication.py (hunk 12 above)
Comment (bmbouter) [reply pulpcore/app/serializers/publication.py]: It is it alphebatizes the fields.

File: pulpcore/app/serializers/publication.py (hunk 13 above)
Comment (bmbouter) [reply pulpcore/app/serializers/publication.py]: this is one aspect I want to fix some after we merge master/detail. I intend to make this clearer over the next day or two.

**Comment by bmbouter** [comment]
I just ran through all of this testing locally and it worked great. I'm going to merge it so I can start into the splitting of `Distribution` into `RepositoryVersionDistribution` and `PublicationDistribution`.
